spinand.c

```diff
@@ -0,0 +1,279 @@
+	if (ret) {
+		msg_cerr("GET FEATURE failed!\n");
+		errno = ret;
+		return 0;
+	}
+
+	msg_cspew("GET FEATURE 0x%01x returned 0x%01x. ", addr, readarr[time - 1]);
+	return readarr[time - 1];
+}
+
+uint8_t spi_nand_get_feature(struct flashctx *flash, unsigned char addr)
+{
+	return spi_nand_get_feature_multi(flash, addr, 1);
+}
+
+int spi_nand_set_feature(struct flashctx *flash, unsigned char addr, uint8_t feature)
+{
+	/* Warning: feature will be kept after soft reset! */
+	unsigned char cmd[SPINAND_SET_FEATURE_OUTSIZE] = { SPINAND_SET_FEATURE, addr, feature };
+
+	return spi_send_command(flash, sizeof(cmd), 0, cmd, NULL);
+}
+
+int spi_nand_wait(struct flashctx *flash)
+{
+	uint8_t feature_c0;
+
+	do {
+		feature_c0 = spi_nand_get_feature_multi(flash, 0xc0, 4);
+		if (errno)
+			return errno;
+	} while (feature_c0 & SPINAND_FEATURE_C0_OIP);
+
+	return 0;
+}
+
+/***************************************
+ * READ CYCLE
+ ***************************************/
+
+/* Typical NAND read process
+ *   1. Read Page (prepare data to the cache)
+ *   2. Get Feature (wait for data cache)
+ *   3. Read Cache
+ */
+
+static int spi_nand_read_page(struct flashctx *flash, unsigned int row_addr)
+{
+	uint8_t cmd[1 + SPINAND_ROW_ADDR_LEN] = { SPINAND_READ_PAGE, };
+
+	int addr_len = spi_nand_prepare_row_address(cmd, row_addr);
+
+	return spi_send_command(flash, 1 + addr_len, 0, cmd, NULL);
+}
+
+static int spi_nand_wait_for_page(struct flashctx *flash, unsigned int row_addr)
+{
+	int ret;
+
+	ret = spi_nand_read_page(flash, row_addr);
+	if (ret)
+		return ret;
+	return spi_nand_wait(flash);
+}
+
+static int spi_nand_read_cache(struct flashctx *flash, unsigned int column_addr,
+                               uint8_t *bytes, unsigned int len)
+{
+	uint8_t cmd[1 + SPINAND_COLUMN_ADDR_LEN] = { SPINAND_READ_CACHE, };
+
+	int addr_len = spi_nand_prepare_column_address(cmd, column_addr);
+
+	/* Send Read */
+	return spi_send_command(flash, 1 + addr_len + DUMMY_BYTE, len, cmd, bytes);
+}
+
+static int spi_nand_inpage_read(struct flashctx *flash, unsigned int row_addr, unsigned int column_addr,
+                                uint8_t *bytes, unsigned int len, unsigned int chunksize)
+{
+	/* Read data within one page */
+	int ret;
+
+	if (chunksize == 0)
+		chunksize = len;
+	ret = spi_nand_wait_for_page(flash, row_addr);
+	if (ret)
+		return ret;
+
+	while (len) {
+		unsigned int data_to_read = min(chunksize, len);
+		ret = spi_nand_read_cache(flash, column_addr, bytes, data_to_read);
+		if (ret)
+			return ret;
+		len -= data_to_read;
+		column_addr += data_to_read;
+		bytes += data_to_read;
+		printf("page %d\n",len);
+	}
+
+	return 0;
+}
+
+static inline unsigned int ilog2(unsigned int x) {
+	int res = 0;
+	while (x >>= 1)
+		++res;
+	return res;
+}
+
+int spi_nand_read_chunked(struct flashctx *flash, uint8_t *buf, unsigned int start,
+                          unsigned int len, unsigned int chunksize)
```

Comment:
My biggest issue with this function is that it sends a read page to cache command for every call to `spi_nand_inpage_read`. This cache read command is followed by a `spi_nand_read_cache`. Some devices only read back 16 bytes at a time. In these cases, an additional 131072-1 commands are issued that do not need to be issued.